src/Filter/Filter.php

```diff
@@ -0,0 +1,193 @@
+<?php
+
+declare(strict_types=1);
+
+namespace swentel\nostr\Filter;
+
+use swentel\nostr\FilterInterface;
+
+class Filter implements FilterInterface
+{
+    /**
+     * A list of event ids
+     */
+    public array $id;
+
+    /**
+     * A list of lowercase pubkeys, the pubkey of an event must be one of these
+     */
+    public array $authors;
+
+    /**
+     * A list of a kind numbers
+     */
+    public array $kinds;
+
+    /**
+     * A list of #e tag values (list of event ids)
+     */
+    public array $etags;
+
+    /**
+     * A list of #p tag values (list of pubkeys).
+     */
+    public array $ptags;
+
+    /**
+     * An integer unix timestamp in seconds, events must be newer than this to pass
+     */
+    public int $since;
+
+    /**
+     * An integer unix timestamp in seconds, events must be older than this to pass 
+     */
+    public int $until;
+
+    /**
+     * Maximum number of events relays SHOULD return in the initial query
+     */
+    public int $limit;
+
+    /**
+     * Set the authors for the Filter object.
+     *
+     * @param array $pubkey The array of authors to set.
+     */
+    public function setAuthors(array $pubkeys): static
+    {
+        foreach($pubkeys as $key){
+            if(!$this->isLowercaseHex($key)){
+                throw new \RuntimeException("Author pubkeys must be an array of 64-character lowercase hex values");
+            }
+        }
+        $this->authors = $pubkeys;
+        return $this;
+    }
+
+    /**
+     * Set the kinds for the Filter object.
+     *
+     * @param array $kinds The array of kinds to set.
+     */
+    public function setKinds(array $kinds): static
+    {
+        $this->kinds = $kinds;
+        return $this;
+    }
+
+    /**
+     * Set the #e tag for the Filter object.
+     *
+     * @param array $etag The array of tag to set.
+     */
+    public function setETag(array $etags): static
+    {
+        foreach($etags as $tag){
+            if(!$this->isLowercaseHex($tag)){
+                throw new \RuntimeException("#e tags must be an array of 64-character lowercase hex values");
+            }
+        }
+        $this->etags = $etags;
+        return $this;
+    }
+
+    /**
+     * Set the #p tag for the Filter object.
+     *
+     * @param array $ptag The array of tag to set.
+     */
+    public function setPTag(array $ptags): static
```

Comment:
I would propose to change the function name to setLowercasePTags as there also can be a setUppercasePTags

Reply:
Good point. I also used isLowercaseHex() for setting #e tags so might be useful to also change setETag() to setLowercaseETags